resourcemanager/pool/spool/spool_test.go

```diff
@@ -0,0 +1,199 @@
+	p.Tune(8)
+	time.Sleep(500 * time.Millisecond)
+	if n := p.Running(); n != 8 {
+		t.Errorf("expect 8 workers running, but got %d", n)
+	}
+	// test pool tune scale down
+	p.Tune(2)
+	for i := 0; i < 6; i++ {
+		c <- struct{}{}
+	}
+	time.Sleep(100 * time.Millisecond)
+	require.Equal(t, 2, p.Running())
+	for i := 0; i < 2; i++ {
+		c <- struct{}{}
+	}
+	time.Sleep(100 * time.Millisecond)
+	require.Equal(t, 0, p.Running())
+
+	// test with RunWithConcurrency
+	var cnt atomic.Int32
+	workerFn := func() {
+		cnt.Add(1)
+	}
+	fnChan := make(chan func(), 10)
+	wg.Wait()
+	err := p.RunWithConcurrency(fnChan, 2)
+	require.NoError(t, err)
+	require.Equal(t, 2, p.Running())
+	for i := 0; i < 10; i++ {
+		fnChan <- workerFn
+	}
+	time.Sleep(100 * time.Millisecond)
+	require.Equal(t, int32(10), cnt.Load())
+	require.Equal(t, 2, p.Running())
+	close(fnChan)
+	time.Sleep(100 * time.Microsecond)
+	require.Equal(t, 0, p.Running())
+	p.ReleaseAndWait()
+}
+
+func TestRunOverload(t *testing.T) {
+	var stop atomic.Bool
+	longRunningFunc := func() {
+		for {
+			if stop.Load() {
+				break
+			}
+			runtime.Gosched()
+		}
+	}
+	poolSize := 10
+	p, err := NewPool("TestMaxBlockingSubmit", int32(poolSize), util.UNKNOWN, WithBlocking(false))
+	require.NoErrorf(t, err, "create TimingPool failed: %v", err)
+	defer p.ReleaseAndWait()
+	defer stop.Store(true)
+	for i := 0; i < poolSize-1; i++ {
+		require.NoError(t, p.Run(longRunningFunc), "submit when pool is not full shouldn't return error")
+	}
+	// p is full now.
+	require.NoError(t, p.Run(longRunningFunc), "submit when pool is not full shouldn't return error")
+	require.EqualError(t, p.Run(demoFunc), pool.ErrPoolOverload.Error(),
+		"blocking submit when pool reach max blocking submit should return ErrPoolOverload")
+}
+
+func TestRunWithNotEnough(t *testing.T) {
+	var stop atomic.Bool
+	fnChan := make(chan func(), 10)
+	poolSize := 10
+	p, err := NewPool("TestRunWithNotEnough", int32(poolSize), util.UNKNOWN, WithBlocking(false))
+	require.NoErrorf(t, err, "create TimingPool failed: %v", err)
+	defer p.ReleaseAndWait()
+	defer stop.Store(true)
+	require.NoError(t, p.RunWithConcurrency(fnChan, uint32(poolSize+100)), "submit when pool is not full shouldn't return error")
+	require.Equal(t, 10, p.Running())
+	require.Error(t, p.RunWithConcurrency(fnChan, 1))
+	require.Error(t, p.Run(func() {}))
+	close(fnChan)
+	time.Sleep(1 * time.Second)
+	require.Equal(t, 0, p.Running())
+}
+
+func TestRunWithNotEnough2(t *testing.T) {
+	fnChan := make(chan func(), 10)
+	var cnt atomic.Int32
+	fn := func() {
+		cnt.Add(1)
+	}
+	p, err := NewPool("TestRunWithNotEnough2", int32(1), util.UNKNOWN, WithBlocking(false))
+	require.NoErrorf(t, err, "create TimingPool failed: %v", err)
+	defer p.ReleaseAndWait()
+	require.NoError(t, p.RunWithConcurrency(fnChan, 2), "submit when pool is not full shouldn't return error")
+	require.Equal(t, 1, p.Running())
+	require.Error(t, p.RunWithConcurrency(fnChan, 1))
+	require.Error(t, p.Run(func() {}))
```

Comment:
These assertions are unstable because the order is decided by the go runtime.

Reply:
No, It is stable. because it tests with logic.

Reply:
OK